### DISCUSSION TITLE
Use secret manager connection for Airflow Metastore DB

### DISCUSSION THREAD

**Discussion by rguillome** [apache/airflow#37328] [Q&A]
Hi,

I have tried to get the airflow db connection config to the metastore db from an aws secret manager key. But in my secret key, infos are stored with JSON format. From `connection.py` I saw that only config elements are read to retrieve an URI (with a valid postgresql/mysql or sqlite URI). Instead I would like to have a connection read from the key provided by the `sql_alchemy_conn_secret` config parameter. In the aws provider, only the method `get_conn_value` read json and transform the dict into a valid URI.
Is there any plans to get the metastore db configuration as in a SecretHook for DAG ? I could participate in the achievement of this feature because today I had to make my own `_cmd` function to build the URI or to make an aws function building the correct secret (like a db uri) aside the existing one used by other services.

Regards,
Guillaume

**Comment by Taragolis** [comment]
Configurations which retrieved from the Alternate Secret Backend (prefixed by _secret) use value as is. I another word [sql_alchemy_conn](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#sql-alchemy-conn) should return values in valid SQLAlchemy URI format regardless of the source: Stored in configurations, retrieved from command or from secret backend.

> only the method get_conn_value read json and transform the dict into a valid URI.

Secrets Backends do not know anything about for which purpose this configuration would be use, it could be `sql_alchemy_conn`, `sql_alchemy_engine_args`, `remote_task_handler_kwargs`, `fernet_key`, `secret_key`, etc. So it pretty hard to implement this on it's side.

> Is there any plans to get the metastore db configuration as in a SecretHook for DAG

As far as i know, there is no such of plan. However Airflow developed by the community and there is 2800~ contributors at that moment. So it always the option to make a PR for implement this feature.

**Reply by rguillome** [reply]
Hi, 

Thanks for your answer. Your point about Secret Backends not knowing anything about which kind of secret a config parameter's got is totally right and that's why I asked the question about the best way to change this a little bit for the metastore connection.

Indeed, in the AWS `SecretsManagerBackend` class when used in DAGs the `get_conn_value` (which is called first by the `get_connection`) manage the json extraction of the secret value to provide a valid URI. I wonder if we could call the this function if the key is found with the `connection_prefix` value of the alternate secret manager first and as we have for the ENV var vs config hierarchy a sort of inspection with ordre : first connection/config even then variables. 
Should I'll check first that the config parameter is from the `database` section and maybe even limit that to the `sql_alchemy_conn` config ?

**Reply by Taragolis** [reply]
May I ask a question? Is it reals a problem problem to store string value into AWS Secret Manager rather than json for this config?
This seems much easier if compare to rewrite Airflow Configuration logic for this case.

**Reply by rguillome** [reply]
Yes It's a _little_ problem since my AWS Secret is rotating so I have to supply a lambda function to manage the rotation and its code comes from a template provided by Amazon from the standard RDS secret representation (with a Json dict).
Indeed I could write a custom to DB URI string rotation but I will move away from the habit of all the other products of my company.
Instead using a standard AWS Secret manager RDS connection configuration would be better.

**Reply by potiuk** [reply]
You can write custom Secret's Manager implementation (extend the AWS one) to unpack Json if the key is about connection. This is absolutely first-class citizen of Airflow to allow you to extend existing classes and roll your own extensions to make something that is needed especially for you. And once you test it, you can event contribute it back to the AWS secrets manager if this is something that we could see as something reusable between other AWS users.

How does it sound as a way forward you could follow @rguillome ?

**Reply by Taragolis** [reply]
> Amazon from the standard RDS secret representation (with a Json dict)

Technically no one stop use in multi-cloud environment e.g. store configs AWS Secrets Manager config, but retrieve it for GCP or Azure, or on-prem. 

Construct valid SQLAlchemy URI it is not straightforward and better keep this tasks for the end users. Some one might required to pass additional parameters through URI and resolve this cases might be a pain.

**Reply by Taragolis** [reply]
>  And once you test it, you can event contribute it back to the AWS secrets manager if this is something that we could see as something reusable between other AWS users.

This is pretty hard to implements, e.g. `sql_alchemy_conn_secret` will contain `awesome-config`. Every secrets backend [community](https://airflow.apache.org/docs/apache-airflow-providers/core-extensions/secrets-backends.html) or user defined will retrieve only `awesome-config`.

It could be a hacky way to get mapping for secrets into the provider by reading airflow configuration. But it also a dangerous due to the fact that secrets backends initialised before entire configuration and there is a chance to accidentally break configurations mechanism. Unfortunetly this part difficult to test in the unit tests because configurations already loaded, and we might only patch already created configs.

**Reply by rguillome** [reply]
> You can write custom Secret's Manager implementation [...] AWS users.
> How does it sound as a way forward you could follow @rguillome ?

Yes I can write a custom Secret's Manager implementation and in fact, the mapping JSON > URI is already made in the `deserialize_connection` of the Secret Backend base implementation. I must only call this method in the first place of the `get_config` on the aws implementtion. Hovewer it sounds a bit hacky because getting a config should not mask / encapsulate a connection building specific code in terms of responsability.
I seems to be the best consensus today.

**Reply by rguillome** [reply]
> Technically no one stop use in multi-cloud environment e.g. store configs AWS Secrets Manager config, but retrieve it for GCP or > Azure, or on-prem.

That's why (from my point of view), the actual mapping is made in the specific AWS Secret backend.

> Construct valid SQLAlchemy URI it is not straightforward and better keep this tasks for the end users. Some one might required to > pass additional parameters through URI and resolve this cases might be a pain.

I though the `extra` key in the JSON string of the secret value was made for ?

> This is pretty hard to implements, e.g. sql_alchemy_conn_secret will contain awesome-config. Every secrets backend [community](https://airflow.apache.org/docs/apache-airflow-providers/core-extensions/secrets-backends.html) or user defined will retrieve only awesome-config.

I'm not sure to get that point. That's already the idea and the actual AWS Secret Manager backend implementation, retrieve the JSON, standarized it, return to  the Base implementation, which use the Core `connection.py` to map an actual Connection. So basically I could make the JSON to URI transform in there (instead of having it in my custom rotating secret function)

> It could be a hacky way to get mapping for secrets into the provider by reading airflow configuration. But it also a dangerous due to the fact that secrets backends initialised before entire configuration and there is a chance to accidentally break configurations mechanism.

Ok that's the actual reason and I get that point. So currently, we already call the Secret backend in the configuration part to get this config. What I suggest is to allow the get_config to have a first part of retrieving a connection-like behaviour. But I should imply having the standardized JSON to URI mapping outside the `Connection.from_json` method ...

In resumé, I could to most of this outside Airflow with custom lambda managing my secret as a database connection URI but in term of exploitation and maintenance, I prefer having it in the aws provider extension. The complexity is already there since the current implementation allows the metastore database configuration to call its `get_config` method

**Reply by Taragolis** [reply]
> That's why (from my point of view), the actual mapping is made in the specific AWS Secret backend.

Still think that without changes in core that would be fragile solution. But if you want, you might raise a PR with your proposal

> I though the extra key in the JSON string of the secret value was made for ?

> But I should imply having the standardized JSON to URI mapping outside the Connection.from_json method ...

Did you mean Postgres Connection from [apache-airflow-providers-postgres](https://airflow.apache.org/docs/apache-airflow-providers-postgres/stable/connections/postgres.html) ? Is so 
1. It only use for operators/hooks but not for 
2. Except some simple cases it build wrong URI
3. There is no requirement to install Postgres Provider for use Postgres as Airflow DB, all you need is psycopg2
4. Configurations and Secrets backends  the part where we try to resolve classical chicken and egg problem and in this part chance that Circular Error Happen pretty high. One of them import `airflow.models.connection.Connection`, see: https://github.com/apache/airflow/pull/26784

**Comment by potiuk** [comment]
> Yes I can write a custom Secret's Manager implementation and in fact, the mapping JSON > URI is already made in the deserialize_connection
> I though the extra key in the JSON string of the secret value was made for ?

This is a common misconception. The connections URIs that Airflow Uses are different standard/format than SQLAlchemy URIs - and you neeed the SQLALCHEMY URI not AIRFLOW URI for metadata DB.

**Reply by rguillome** [reply]
Ok thanks I didn't know that ! So definitly my best choices are to extend the actual aws custom secret backends or provide a secret directly in the suitable sql_alchemy form (and with rotation expectation the secret rotation lambda).
Regarding organisation simplicity my current choice is to provide an extend of the actual aws custom secret backend

**Reply by potiuk** [reply]
Yep.
